controllers/volumereplicationgroup_webhook.go

```diff
@@ -0,0 +1,141 @@
+	log := v.log
+
+	vrg := &ramen.VolumeReplicationGroup{}
+	if err := v.decoder.Decode(req, vrg); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	vrgNamespacedName := types.NamespacedName{
+		Namespace: vrg.Namespace,
+		Name:      vrg.Name,
+	}
+	log = log.WithValues("vrg", vrgNamespacedName.String())
+
+	if vrg.Spec.KubeObjectProtection == nil ||
+		vrg.Spec.KubeObjectProtection.RecipeRef == nil {
+		return admission.Allowed("recipe not specified")
+	}
+
+	recipeNamespacedName := types.NamespacedName{
+		Namespace: vrg.Spec.KubeObjectProtection.RecipeRef.Namespace,
+		Name:      vrg.Spec.KubeObjectProtection.RecipeRef.Name,
+	}
+
+	recipe := &recipe.Recipe{}
+	if err := v.client.Get(ctx, recipeNamespacedName, recipe); err != nil {
+		return admission.Errored(http.StatusInternalServerError,
+			fmt.Errorf("recipe %#v retrieval error: %w", recipeNamespacedName, err))
+	}
+
+	if err := RecipeParametersExpand(recipe, vrg.Spec.KubeObjectProtection.RecipeParameters, log); err != nil {
+		return admission.Errored(http.StatusInternalServerError,
+			fmt.Errorf("recipe %+v vrg %+v parameter expansion error: %w", recipe, vrg, err))
+	}
+
+	return v.recipeNamespacesValidate(ctx, req, recipe)
+}
+
+func (v *vrgValidator) recipeNamespacesValidate(ctx context.Context, req admission.Request, recipe *recipe.Recipe,
+) admission.Response {
+	namespaceNames := make(sets.Set[string], 0)
+
+	if recipe.Spec.Volumes != nil {
+		namespaceNames.Insert(recipe.Spec.Volumes.IncludedNamespaces...)
+	}
+
+	for _, group := range recipe.Spec.Groups {
+		namespaceNames.Insert(group.IncludedNamespaces...)
+	}
+
+	for _, namespaceName := range namespaceNames.UnsortedList() {
+		accessReview := vrgCreateAccessReview(req.UserInfo, namespaceName)
+		if err := v.client.Create(ctx, accessReview); err != nil {
+			return admission.Errored(http.StatusInternalServerError,
+				fmt.Errorf("namespace %v access review creation error: %w",
+					namespaceName, err))
+		}
+
+		if !accessReview.Status.Allowed {
+			return admission.Denied(
+				fmt.Sprintf("%v %v creation access denied: %+v",
+					vrgKindName,
+					accessReview.Spec.ResourceAttributes.Namespace,
+					accessReview.Status,
+				),
+			)
+		}
+	}
+
+	return admission.Allowed("")
+}
+
+func vrgCreateAccessReview(userInfo authentication.UserInfo, namespaceName string,
+) *authorization.SubjectAccessReview {
+	extra := make(map[string]authorization.ExtraValue, len(userInfo.Extra))
+	for key, value := range userInfo.Extra {
+		extra[key] = authorization.ExtraValue(value)
+	}
+
+	return &authorization.SubjectAccessReview{
+		Spec: authorization.SubjectAccessReviewSpec{
+			ResourceAttributes: &authorization.ResourceAttributes{
+				Namespace: namespaceName,
+				Verb:      "create",
+				Group:     ramen.GroupVersion.Group,
+				Version:   ramen.GroupVersion.Version,
+				Resource:  vrgResourceName,
+			},
+			User:   userInfo.Username,
```

Comment:
If I understand this right, this will validate the user creating the VRG, which in the case of the managed cluster via OCM/ACM would be via the klusterlet, and that has RBAC to create VRG in all namespaces.

What hence would happen is that when DRPC rolls this out to a ManagedCluster, it would always be allowed to be created. Which defeats the purpose of the check and the security it intends to provide.

The check has to happen in such cases at the hub as a result.

Reply:
For MVP, multi-namespace recipes are only allowed for VRGs is in an "admin" namespace.